Framework/DataHandling/src/SetSample.cpp

```diff
@@ -405,6 +588,109 @@ const Geometry::SampleEnvironment *SetSample::setSampleEnvironment(
   return &(exptInfo.sample().getEnvironment());
 }
 
+/**
+ * Set the requested sample environment from shape XML string
+ * @param exptInfo A reference to the ExperimentInfo to receive the environment
+ * @param canGeomArgs The dictionary of flags for the environment
+ * @param canMaterialArgs The dictionary of material parameters
+ * @return A pointer to the new sample environment
+ */
+const Geometry::SampleEnvironment *SetSample::setSampleEnvironmentFromXML(
+    API::ExperimentInfo &exptInfo,
+    const Kernel::PropertyManager_const_sptr &canGeomArgs,
+    const Kernel::PropertyManager_const_sptr &canMaterialArgs) {
+  const auto refFrame = exptInfo.getInstrument()->getReferenceFrame();
+  const auto xml = tryCreateXMLFromArgsOnly(*canGeomArgs, *refFrame);
+  if (!xml.empty()) {
+    ShapeFactory sFactory;
+    // Create the object
+    auto shape = sFactory.createShape(xml);
+    if (shape->hasValidShape()) {
+      if (canMaterialArgs) {
+        ReadMaterial::MaterialParameters materialParams;
+        setMaterial(materialParams, *canMaterialArgs);
+        if (materialParams.sampleVolume <= 0.) {
+          materialParams.sampleVolume = shape->volume() * CUBIC_METRE_TO_CM;
+        }
+        ReadMaterial reader;
+        reader.setMaterialParameters(materialParams);
+        auto canMaterial = reader.buildMaterial();
+        shape->setMaterial(*canMaterial);
+      }
+      const SampleEnvironment se("unnamed", std::make_shared<Container>(shape));
+      exptInfo.mutableSample().setEnvironment(
+          std::make_unique<SampleEnvironment>(se));
+    }
+  }
+  return &(exptInfo.sample().getEnvironment());
+}
+
+/**
+ * @brief SetSample::setMaterial Configures a material from the parameters
+ * @param materialParams : output material parameters object
+ * @param materialArgs : input material arguments
+ */
+void SetSample::setMaterial(ReadMaterial::MaterialParameters &materialParams,
+                            const Kernel::PropertyManager &materialArgs) {
+  if (materialArgs.existsProperty("ChemicalFormula")) {
+    materialParams.chemicalSymbol =
+        materialArgs.getPropertyValue("ChemicalFormula");
+  }
+  if (materialArgs.existsProperty("AtomicNumber")) {
+    materialParams.atomicNumber = materialArgs.getProperty("AtomicNumber");
+  }
+  if (materialArgs.existsProperty("MassNumber")) {
+    materialParams.massNumber = materialArgs.getProperty("MassNumber");
+  }
+  if (materialArgs.existsProperty("SampleNumberDensity")) {
```

Comment:
That's a good point, ideally these properties should say MassDensity and NumberDensity, since it does not matter it's about the sample or anything else. However I did not change this for not breaking the API. One could prefix them with Container, but that will need a tweak in the validation I guess.

Reply:
If you kept the logic for "SampleNumberDensity" and added the following lines in setMaterial I don't think there would be any negative impact on existing scripts that call SetSample:
```
  if (materialArgs.existsProperty("ContainerNumberDensity")) { 
    materialParams.sampleNumberDensity = 
        materialArgs.getProperty("ContainerNumberDensity"); 
  }
```
...or you could alternatively support "NumberDensity" if you prefer.

You could also change the field name on the materialParams struct to "numberDensity" but this would require a few changes to other units that use this struct so this is probably lower priority.